File: csharp/16_async_await.cs
// ============================================================
// Topic:  Async / Await  (FOCUS: Task, await, Task.WhenAll, delays, errors)
// Summary: Non-blocking async methods — method recall.
// Run:     dotnet script csharp/16_async_await.cs
//
// KEY DIFFERENCES:
//   C#:     async Task / async Task<T>, await, Task.WhenAll
//   Python: async def, await, asyncio.gather
//   TS/JS:  async function, await, Promise.all
//   Rust:   async fn, .await, tokio::join!
//   Java:   CompletableFuture.supplyAsync — no async/await keyword until recent patterns
// ============================================================

using System;
using System.Threading.Tasks;

// ----- 1. async + await — basic async method -----
async Task<string> SayHello()
{
    await Task.Delay(10);                       // yield     | Python: await asyncio.sleep(0.01) | TS: await delay(10)
    return "hello";
}
Console.WriteLine(await SayHello());            // hello     | TS: await sayHello() | Python: await say_hello()

// ----- 2. Sequential awaits — one after another -----
async Task<string> Step1() { await Task.Delay(10); return "step1"; }
async Task<string> Step2() { await Task.Delay(10); return "step2"; }
Console.WriteLine(await Step1());               // step1     | first task finishes before second starts
Console.WriteLine(await Step2());               // step2     | Python: s1 = await step1(); s2 = await step2()

// ----- 3. Concurrent Task.WhenAll — run in parallel -----
async Task<string> WorkA() { await Task.Delay(10); return "A"; }
async Task<string> WorkB() { await Task.Delay(10); return "B"; }
string[] both = await Task.WhenAll(WorkA(), WorkB());
Console.WriteLine($"[{string.Join(", ", both)}]"); // [A, B] | Python: await asyncio.gather(work_a(), work_b()) | TS: Promise.all([workA(), workB()])

// ----- 4. Task.Delay — non-blocking pause -----
async Task DelayDemo()
{
    await Task.Delay(10);                       // delay 10ms | Python: await asyncio.sleep(0.01) | TS: await new Promise(r => setTimeout(r, 10))
    Console.WriteLine("done");                  // done
}
await DelayDemo();

// ----- 5. try/catch in async — exception handling -----
async Task Boom()
{
    await Task.Delay(10);
    throw new Exception("boom");
}
try
{
    await Boom();
}
catch (Exception ex)
{
    Console.WriteLine($"caught: {ex.Message}"); // caught: boom | TS: try { await boom() } catch (e) | Python: try/except around await
}

// ----- 6. Chaining async results -----
async Task<int> GetValue() { await Task.Delay(10); return 42; }
async Task<string> Chain()
{
    int n = await GetValue();                   // await intermediate | Python: n = await get_value()
    return $"chained: {n}";
}
Console.WriteLine(await Chain());               // chained: 42 | pipe async results through await

// Cross-language note:
//   C# async/await compiles to a state machine — same idea as TS/JS Promises under the hood.
//   Task.WhenAll = Python asyncio.gather / JS Promise.all — tasks run concurrently, results keep input order.
//   Sequential await blocks only the current async method, not the thread (unlike Thread.Sleep).
//   dotnet script supports top-level await; a normal .csproj needs an async Main or host builder.

File: cpp/16_async_await.cpp
// ============================================================
// Topic:  Async / Await  (FOCUS: std::async, std::future, .get(), delays)
// Summary: Concurrent tasks via futures — sequential vs parallel execution.
// Run:     g++ -std=c++17 cpp/16_async_await.cpp -o cpp/16_async_await && cpp/16_async_await
//
// KEY DIFFERENCES:
//   C++:     std::async + std::future + .get()  (C++17 — no async/await keyword)
//   C#:      async Task / await
//   Python:  async def / await / asyncio.gather
//   TS/JS:   async function / await / Promise.all
//   Rust:    async fn / .await / tokio::join!
//   Java:    CompletableFuture.supplyAsync
//   C++20:   co_await coroutines — needs -std=c++20 and custom awaitables
// ============================================================

#include <chrono>
#include <future>
#include <iostream>
#include <stdexcept>
#include <string>
#include <thread>

std::string do_say_hello() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "hello";
}

std::string do_step1() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "step1";
}

std::string do_step2() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "step2";
}

std::string do_work_a() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "A";
}

std::string do_work_b() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "B";
}

int do_boom() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    throw std::runtime_error("boom");
}

int do_get_value() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return 42;
}

std::string do_delay_demo() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return "done";
}

std::string do_chain() {
    int n = do_get_value();
    return "chained: " + std::to_string(n);
}

int main() {
    // ----- 1. std::async + future.get() — like await -----
    std::cout << std::async(std::launch::async, do_say_hello).get() << std::endl;
    // Output: hello

    // ----- 2. Sequential .get() — one after another -----
    std::cout << std::async(std::launch::async, do_step1).get() << std::endl;
    // Output: step1
    std::cout << std::async(std::launch::async, do_step2).get() << std::endl;
    // Output: step2

    // ----- 3. Concurrent — launch both, then collect results -----
    auto fa = std::async(std::launch::async, do_work_a);
    auto fb = std::async(std::launch::async, do_work_b);
    std::cout << "[" << fa.get() << ", " << fb.get() << "]" << std::endl;
    // Output: [A, B]

    // ----- 4. sleep_for — delay inside async task -----
    std::cout << std::async(std::launch::async, do_delay_demo).get() << std::endl;
    // Output: done

    // ----- 5. try/catch around .get() — exception propagation -----
    try {
        std::async(std::launch::async, do_boom).get();
    } catch (const std::exception& ex) {
        std::cout << "caught: " << ex.what() << std::endl;
    }
    // Output: caught: boom

    // ----- 6. Chaining async results -----
    std::cout << std::async(std::launch::async, do_chain).get() << std::endl;
    // Output: chained: 42

    // Cross-language note:
    //   C++ has no async/await keyword in C++17 — .get() on std::future is the blocking "await".
    //   Launch tasks first, then .get() all of them for parallelism (like Promise.all / gather).
    //   C++20 co_await + coroutines mirror async/await syntax but need custom promise types.
    return 0;
}